### PR TITLE
SI-8126 add -Xsource option to let (parts of) the compiler run on older code

### PR DESCRIPTION
**Pull request by huitseeker**
This adds a `-source` flag and puts various 2.10 source-level
compatibility-breaking fixes under it (so that the fixed codepath
doesn't run with -source:2.10). 

The last commit adds another (early) setting (`-Yrelax-variance-checks`)
for the specific checks brought by SI-6566, fixed in a419799f87. This
is needed to bypass the stricter (and correct) variance checks done in
`reflect/internal/Variances`. `-Y-relax-variance-checks`, contrarily to
`-source`, doesn't depend on the `nsc.settings.ScalaVersion` class. This
option is turned on by `-source:2.10`

This has been tested by compiling the 2.10 scala library (using the
compiler in this PR , with "`-source:2.10 -Ystop-after:typer`") using
this project: https://github.com/huitseeker/scala-library

review by @dragos or @retronym.

### CODE REVIEW

**Comment by huitseeker** [comment]
I higly advise looking at diffs for `ScalaSettings.scala` without whitespace (i.e. adding a '?w=1' to the github URL you browse to).

**Comment by adriaanm** [comment]
Very cool! I think a boolean flag like isScala211 would be faster to test, I didn't look very closely to see whether these are hotspots in the typechecker, though.

**Comment by huitseeker** [comment]
@adriaanm A binary 2.11 flag prevents passing a finer threshold than 2.11, such as `-source:2.11.42`, and it's harder to maintain, but it does simplify the PR and yields faster type checking. To err on the side without premature optimization, I updated with your suggestion. /cc @retronym @dragos, who might want to chime in. The old branch is still visible [there](https://github.com/huitseeker/scala/tree/issue/sourceFlag). Updated [the test project](https://github.com/huitseeker/scala-library) too.

**Comment by dragos** [comment]
I'd prefer to have finer-grained levels, since changes in the type-checker might appear even in minor releases (fix an unsoundness bug, or see [this case](https://groups.google.com/forum/#!searchin/scala-internals/standstill/scala-internals/8rZeCeiTYDo/g37r5Lex7g4J)). It would be very confusing to have errors in the presentation compiler that the build compiler won't issue. We could map versions to ints, and do an int comparison instead, wdyt?

**Comment by retronym** [comment]
Fine grained levels are good at the command line, but we should centralize the common checks into helper functions in settings to avoid litterring duplicated version strings about the place. That also gives us a spot to cache those. 

I also wonder if this option should be -Y for now. Is it really ready to show up in scala -help?

**Comment by dragos** [comment]
Sure, they should be centralized, probably having final vals or supported releases, a sort of poor man's enumeration.

If you prefer to hide this option from users, I don't mind not showing it in the scalac -help page.. how about -X? I thought -Y is private/debugging options, while -X is experimental. Either way is fine with me.

**Comment by retronym** [comment]
`-X` is probably suitable. This option sort of generalizes the existing option `-Xfuture`.

**Comment by retronym** [comment]
I also tend to think we should not do anything more fine grained than `2.10` / `2.11`. This is in line with Java's approach.

If we want to deliver a bug fix in, say, 2.11.3 that reports true errors that were not reported in 2.11.0, we could either:
- rethink, and defer it to 2.12.0
- put it behind `-Xsource:2.12` in 2.11.3.

**Comment by adriaanm** [comment]
Thanks for the fast turnaround, @huitseeker, but I meant my suggestion as an internal flag that checks the command line setting once and can then be checked quickly during type checking. I like having a -Xsource flag. I'm with Jason, though, that we should restrict it to epoch.major. Minor scala versions aren't allowed to be source incompatible, so the case @dragos mentioned to motivate finer-grained versions should be treated as a bug.

**Comment by huitseeker** [comment]
Okay, I updated this, also removing the last commit flag-masking the fix of SI-6566, since @retronym reminded me yesterday this is done in refchecks (we're aiming for `-Ystop-after:typer`).

**Comment by huitseeker** [comment]
BTW, I'm going to update this again, but the kitteh complaining is irrelevant to the PR (it complains about deployment pbs)

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Macros pose further challenges to this approach: https://scala-ide-portfolio.assembla.com/spaces/scala-ide/tickets/1001449?comment=431750413#comment:431750413

We can at least make the macro expansion engine in 2.11 backwards compatible with the format that macro metadata is stored (@xeno-by will do this before 2.11)

But even then, we won't have binary compatibility of the reflection API from 2.11, so its likely that macro expansion will result in linkage errors.

This might be a good time for the new option `-Ymacro-expand:discard-whitebox-only` discussed in https://github.com/scala/scala/pull/3287. This would let us exploit the black-boxity of some macros and avoid expansion altogether. But for full functionality (expansion of macros to compute result types of white-box macros and to show macro-specific errors/warnings), we would still need to ship an IDE with the 2.10 presentation compiler.

**Comment by retronym** [comment]
Macros in the standard library probably won't suffer from this fate as they go through the "FastTrack" mechanism, so the macro implementations are shipped in the compiler.

**Comment by adriaanm** [comment]
(Note that this needs a rebase.)

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 1]
```diff
@@ -73,45 +73,52 @@ trait ScalaSettings extends AbsScalaSettings
...
-  val mainClass     = StringSetting     ("-Xmain-class", "path", "Class for manifest's Main-Class entry (only useful with -d <jar>)", "")
-  val Xshowcls      = StringSetting     ("-Xshow-class", "class", "Show internal representation of class.", "")
-  val Xshowobj      = StringSetting     ("-Xshow-object", "object", "Show internal representation of object.", "")
-  val showPhases    = BooleanSetting    ("-Xshow-phases", "Print a synopsis of compiler phases.")
-  val sourceReader  = StringSetting     ("-Xsource-reader", "classname", "Specify a custom method for reading source files.", "")
-  val strictInference = BooleanSetting  ("-Xstrict-inference", "Don't infer known-unsound types")
+  val logFreeTerms       = BooleanSetting    ("-Xlog-free-terms", "Print a message when reification creates a free term.")
+  val logFreeTypes       = BooleanSetting    ("-Xlog-free-types", "Print a message when reification resorts to generating a free type.")
+  val maxClassfileName   = IntSetting     ("-Xmax-classfile-name", "Maximum filename length for generated classes", 255, Some((72, 255)), _ => None)
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: Was it intended to align parens?

**Comment by dragos** [comment]
@adriaanm Does this mean there won't be any new compilation errors in a minor release (fixing soundness bugs, for instance, are pushed only in major versions?)

**Comment by retronym** [comment]
We can deliver them in a minor release but hide them under `-Xfuture` or `-Xsource:2.12`.

It is still possible (probable?) that some things will sneak through. Some support from dbuild might be necessary to give us confidence.

**Comment by dragos** [comment]
@retronym, of course, I know there's no perfect software. I just want to understand clearly what we design for. @huitseeker is looking into dbuild support for regression testing this flag.

**Comment by adriaanm** [comment]
@dragos, that's a good point -- we could consider fixing a really really bad soundness issues in minor releases. Normally we don't even touch type inference / implicit search for fear of causing source incompatibilities. So, you could say the bad ones need a source compat flag even more urgently, but I would say this is also supported by @retronym's proposal, where we classify these under the source-level of the next major release.

For example, let's say we discover a major soundness problem in 2.10 (fingers crossed!) and we decide somehow it's important enough to fix in 2.10.x, we could fix it, but disable the new check under -Xsource:2.10. When no -Xsource (or -Xsource:2.11) is given, scalac would reject more programs.

**Comment by dragos** [comment]
Makes sense, thanks for explanations!

**Comment by huitseeker** [comment]
Rebased, re-whitespace-aligned.

**Comment by adriaanm** [comment]
Good to merge as soon as https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/5441/console succeeds

**Comment by adriaanm** [comment]
LGTM

**Comment by adriaanm** [comment]
Build passed. Merging! Nice work, @huitseeker!

**Pull request merged**